### PR TITLE
Add ESM-2 model gradient tests

### PR DESCRIPTION
**Pull request by pstjohn**
Adds gradient tests to ensure the gradients we get from ddp, fsdp2, and nvfsdp are consistent

## Summary by CodeRabbit

- Chores
  - Simplified image builds by switching to standard pip for editable installs.
- Dependencies
  - Updated core ML dependencies and enabled/unpinned select components for better compatibility.
- Tests
  - Added comprehensive distributed-training validation tests comparing strategies across single- and multi-GPU setups with automatic GPU gating and detailed checks.
- Refactor
  - Extracted test data generation into a reusable helper to streamline test setup.

### CODE REVIEW

File: models/esm2/tests/test_distributed_strategies.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+        # get gradients
+        if strategy is Strategy.FSDP2:
+            grads = {name: p.grad.full_tensor() for name, p in model.named_parameters() if p.grad is not None}
+
+        elif strategy is Strategy.DDP:
+            grads = {name: p.grad for name, p in model.module.named_parameters() if p.grad is not None}
+
+        elif strategy is Strategy.MFSDP:
+            grads = {name: p.grad.full_tensor() for name, p in model.module.named_parameters() if p.grad is not None}
```
Comment (pstjohn) [inline models/esm2/tests/test_distributed_strategies.py]: @cspades do you see anything here that would be causing multi-device tests to hang?

File: models/esm2/tests/test_distributed_strategies.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        stderr=subprocess.PIPE,
+        timeout=240,
+    )
+    if result.returncode != 0:
+        print(f"STDOUT:\n{result.stdout}")
+        print(f"STDERR:\n{result.stderr}")
+        pytest.fail(f"Command failed with exit code {result.returncode}")
+
+
+@requires_multi_gpu
```
Comment (jomitchellnv) [inline models/esm2/tests/test_distributed_strategies.py:68]: do we need a more precise decorator that says "requires 2 gpus"?

File: models/esm2/tests/test_distributed_strategies.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+if __name__ == "__main__":
+    import argparse
+    import enum
+    from dataclasses import dataclass, field
+
+    import torch.distributed as dist
+    import transformer_engine.pytorch
+    import transformers
+    from megatron_fsdp.fully_shard import fully_shard as megatron_fsdp_fully_shard
```
Comment (jomitchellnv) [inline models/esm2/tests/test_distributed_strategies.py:104]: so we have to import mcore just to get nvFSDP? yikes

File: models/esm2/tests/test_distributed_strategies.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+        rank: int = field(default_factory=dist.get_rank)
+        local_rank: int = field(default_factory=lambda: int(os.environ["LOCAL_RANK"]))
+        world_size: int = field(default_factory=dist.get_world_size)
+
+        def is_main_process(self) -> bool:
+            """This is the global rank 0 process, to be used for wandb logging, etc."""
+            return self.rank == 0
+
+    def run_forward_backward(use_te: bool, strategy: Strategy, input_data: dict, dist_config: DistributedConfig):
```
Comment (jomitchellnv) [inline models/esm2/tests/test_distributed_strategies.py:136]: so much if else if stuff, maybe just split it out for separate functions at the top of the file?

File: models/esm2/tests/test_distributed_strategies.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        timeout=240,
+    )
+    if result.returncode != 0:
+        print(f"STDOUT:\n{result.stdout}")
+        print(f"STDERR:\n{result.stderr}")
+        pytest.fail(f"Command failed with exit code {result.returncode}")
+
+
+@requires_multi_gpu
+@pytest.mark.parametrize("strategy", ["fsdp2", "mfsdp"])
```
Comment (jomitchellnv) [inline models/esm2/tests/test_distributed_strategies.py]: I always prefer to have more tests than to use parametrize here, that way its clear. It's easier to edit down the line.

File: models/esm2/tests/test_distributed_strategies.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+@pytest.mark.parametrize("backend", ["te", "eager"])
+def test_ddp_vs_fsdp2_single_gpu(strategy, backend):
+    cmd = [
+        "torchrun",
+        "--nproc_per_node=1",
+        os.path.relpath(__file__),
+        "--strategy",
+        strategy,
+    ]
+    if backend == "te":
```
Comment (jomitchellnv) [inline models/esm2/tests/test_distributed_strategies.py:51]: nit: you should just make another test that doesn't use TE.

File: models/esm2/tests/test_distributed_strategies.py (hunk 3 above)
Comment (pstjohn) [reply models/esm2/tests/test_distributed_strategies.py:104]: no, this is `megatron-fsdp`: [pypi.org/project/megatron-fsdp](pypi.org/project/megatron-fsdp)

we're not calling it nvFSDP anymore. It's megatron-fsdp.

File: models/esm2/tests/test_distributed_strategies.py (hunk 5 above)
Comment (pstjohn) [reply models/esm2/tests/test_distributed_strategies.py]: fair; id just need to have multiple test files probably then. These tests are slightly odd in that there's a `__main__` entrypoint we call from the actual pytest tests, since we need to invoke these as subprocesses to ensure they're run with multiple processes

File: models/esm2/tests/test_distributed_strategies.py (hunk 1 above)
Comment (cspades) [reply models/esm2/tests/test_distributed_strategies.py]: Can we change it to this:
```
        elif strategy is Strategy.MFSDP:
            # Because of uneven sharding, we need to manually gather the gradients.
            sharded_grads = [(name, p.grad) for name, p in model.module.named_parameters()]
            grads = {}
            for name, grad in sharded_grads:
                grad_shards = [None] * device_mesh["dp"].size()
                # For FSDP, we are not strided sharding, so gathering across dp_shard_cp is sufficient.
                # For HSDP, we need to first gather across dp_shard_cp, then gather across dp_inter,
                # not the other way around or you'll get wrong zig-zags.
                torch.distributed.all_gather_object(grad_shards, grad, group=device_mesh["dp"].get_group())
                all_valid_shards = [shard for shard in grad_shards if shard is not None]
                # Megatron-FSDP is always sharded across dim=0.
                grads[name] = torch.cat([s.to_local().to(device) for s in all_valid_shards], dim=0)
```
The test will still fail as-is due to our gradient x2 issue, but this code will be useful for the future.

File: models/esm2/tests/test_distributed_strategies.py (hunk 6 above)
Comment (pstjohn) [reply models/esm2/tests/test_distributed_strategies.py:51]: right, as explained above i agree but this test suite is a little different, since we need to call `torchrun` on a file with different argparse commands. idk. nice to keep them together in the same file
